### PR TITLE
feat: receive managerCallbackUrl from job params - overseer (MAPCO-2384)

### PR DESCRIPTION
**Pull request by ronenkapelian**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✔                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

- Change the logic, instead receive the overseerUrl from configuration, now it will receive it from job's parameter.
- Also the name should be more generic and less dedicated so it change to "ManagerCallbackUrl.

### CODE REVIEW

File: MergerService/Utils/JobUtils.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            _jsonSerializerSettings = new JsonSerializerSettings();
+            _jsonSerializerSettings.Converters.Add(new StringEnumConverter());
+
+            _jobManagerUrl = this._configuration.GetConfiguration("TASK", "jobManagerUrl");
+            //TODO: add tracing
+        }
+
+        public MergeJob? GetJob(string jobId)
+        {
+            using (this._activitySource.StartActivity("dequeue task"))
```
Comment (asafMasa) [inline MergerService/Utils/JobUtils.cs:40]: this is a wrong activity for getting a job

File: MergerService/Utils/JobUtils.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                string relativeUri = $"jobs/{jobId}";
+                string url = new Uri(new Uri(_jobManagerUrl), relativeUri).ToString();
+                string? jobData = this._httpClient.GetDataString(url);
+                if (jobData is null)
+                {
+                    this._logger.LogInformation($"Job id:{jobData}, not found");
```
Comment (asafMasa) [inline MergerService/Utils/JobUtils.cs:47]: warn level as it searches for it's own job that isn't exists

File: MergerService/Utils/JobUtils.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                    return null;
+                }
+
+                try
+                {
+                    return JsonConvert.DeserializeObject<MergeJob>(jobData, this._jsonSerializerSettings)!;
+                }
+                catch (Exception e)
+                {
+                    this._logger.LogWarning(e.Message);
```
Comment (asafMasa) [inline MergerService/Utils/JobUtils.cs:57]: i think it should be an error level message

File: MergerService/src/run.cs [hunk 4]
```diff
@@ -190,10 +203,11 @@ public void Start()
                     }
 
                     activatedAny = true;
+
```
Comment (asafMasa) [inline MergerService/src/run.cs:206]: remove extra row

File: MergerService/Utils/TaskUtils.cs [hunk 5]
```diff
@@ -125,9 +126,9 @@ public void UpdateReject(string jobId, string taskId, int attempts, string reaso
                 attempts++;
 
                 // Check if the task should actually fail
-                if (!resettable || attempts == this._maxAttempts)
+                if (managerCallbackUrl is not null && (!resettable || attempts == this._maxAttempts))
```
Comment (asafMasa) [inline MergerService/Utils/TaskUtils.cs:129]: managerCallbackUrl can be null, why did you add it here?

File: MergerService/Utils/TaskUtils.cs [hunk 6]
```diff
@@ -155,8 +156,11 @@ private void UpdateFailed(string jobId, string taskId, int attempts, string reas
                 }
             }
 
+            if (managerCallbackUrl is not null)
+            {
             // Notify overseer on task failure
-            NotifyOnStatusChange(jobId, taskId);
+            NotifyOnStatusChange(jobId, taskId, managerCallbackUrl);
```
Comment (asafMasa) [inline MergerService/Utils/TaskUtils.cs:162]: search for the word "overseer" and replace it with "managerCallback"

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good Job 👍 , please mind my remarks

File: MergerService/Utils/TaskUtils.cs (hunk 5 above)
Comment (ronenkapelian) [reply MergerService/Utils/TaskUtils.cs:129]: because it shouldn't update failed, it uses the callback URL, but it should update the job - for exporter situations.
i consult with Shimon

File: MergerService/Utils/JobUtils.cs (hunk 3 above)
Comment (ronenkapelian) [reply MergerService/Utils/JobUtils.cs:57]: Done

File: MergerService/Utils/JobUtils.cs (hunk 2 above)
Comment (ronenkapelian) [reply MergerService/Utils/JobUtils.cs:47]: done

**Review by asafMasa — APPROVED** [review]
Good Job 🥇

File: MergerService/MergerService.csproj [hunk 7]
```diff
@@ -1,27 +1,23 @@
...
   </ItemGroup>
-
   <ItemGroup>
     <Content Update="appsettings.json">
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>
   </ItemGroup>
-
-</Project>
+</Project>
```
Comment (shimoncohen) [inline MergerService/MergerService.csproj:23]: Add newline at the end of the file.

File: MergerService/Models/Jobs/JobMergeMetadata.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            Console.WriteLine(this.ToString());
+        }
+
+        public override string ToString()
+        {
+            return JsonConvert.SerializeObject(this, this._jsonSerializerSettings)!;
+        }
+    }
+
+}
```
Comment (shimoncohen) [inline MergerService/Models/Jobs/JobMergeMetadata.cs:43]: Add newline at the end of the file.

File: MergerService/Models/Jobs/JobMetadata.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            Console.WriteLine(this.ToString());
+        }
+
+        public override string ToString()
+        {
+            return JsonConvert.SerializeObject(this, this._jsonSerializerSettings)!;
+        }
+    }
+
+}
```
Comment (shimoncohen) [inline MergerService/Models/Jobs/JobMetadata.cs:92]: Add newline at the end of the file.

File: MergerService/Models/Jobs/JobMergeMetadata.cs [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        public JobMergeMetadata(JobMetadata metadata, string[] fileNames,string originDirectory, string layerRelativePath,
+            string managerCallbackUrl)
+        {
+            this.Metadata = metadata;
+            this.FileNames = fileNames;
+            this.OriginDirectory = originDirectory;
+            this.LayerRelativePath = layerRelativePath;
+            this.ManagerCallbackUrl = managerCallbackUrl;
+
```
Comment (shimoncohen) [inline MergerService/Models/Jobs/JobMergeMetadata.cs:26]: Remove empty line.

File: MergerService/Models/Jobs/JobMetadata.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        private JsonSerializerSettings _jsonSerializerSettings;
+
+        public JobMetadata(string id, double rms, string type, int scale, string srsId, string[] region,
+            string[] sensors, string srsName, GeoJson footprint, string productId, string description,
+            string displayPath, string productName, string productType, DateTime creationDate, string producerName,
+            DateTime ingestionDate, DateTime sourceDateEnd, string classification, string productSubType,
+            string productVersion, DateTime sourceDateStart, double maxResolutionDeg, FeatureCollection layerPolygonParts,
+            double maxResolutionMeter, string productBoundingBox, double minHorizontalAccuracyCE90)
+        {
+            this.Id = id;
```
Comment (shimoncohen) [inline MergerService/Models/Jobs/JobMetadata.cs:49]: Nice.

File: MergerService/Utils/TaskUtils.cs [hunk 12]
```diff
@@ -111,11 +109,14 @@ public void UpdateCompletion(string jobId, string taskId)
                 }
             }
 
-            // Update overseer on task completion
-            NotifyOnStatusChange(jobId, taskId);
+            if (managerCallbackUrl is not null)
+            {
+            // Update Manager on task completion
```
Comment (shimoncohen) [inline MergerService/Utils/TaskUtils.cs:114]: Please use proper indentation.

File: MergerService/Utils/TaskUtils.cs [hunk 13]
```diff
@@ -125,9 +126,9 @@ public void UpdateReject(string jobId, string taskId, int attempts, string reaso
                 attempts++;
 
                 // Check if the task should actually fail
-                if (!resettable || attempts == this._maxAttempts)
+                if ((!resettable || attempts == this._maxAttempts))
```
Comment (shimoncohen) [inline MergerService/Utils/TaskUtils.cs:129]: Please remove one set of parentheses.

File: MergerService/Utils/TaskUtils.cs [hunk 14]
```diff
@@ -155,8 +156,11 @@ private void UpdateFailed(string jobId, string taskId, int attempts, string reas
                 }
             }
 
-            // Notify overseer on task failure
-            NotifyOnStatusChange(jobId, taskId);
+            if (managerCallbackUrl is not null)
+            {
+            // Notify Manager on task failure
```
Comment (shimoncohen) [inline MergerService/Utils/TaskUtils.cs:161]: Please use proper indentation.

File: MergerService/src/run.cs [hunk 15]
```diff
@@ -164,18 +167,28 @@ public void Start()
                         continue;
                     }
 
+                    string? managerCallbackUrl = jobUtils.GetJob(task.JobId)?.Parameters.ManagerCallbackUrl;
+                    if (managerCallbackUrl == null)
```
Comment (shimoncohen) [inline MergerService/src/run.cs:171]: Maybe use trenary here for more readable code?

**Review by shimoncohen — APPROVED** [review]
